core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,771 @@
+}
+
+static TEE_Result sec_cipher_aes_get_c_key_len(size_t key_len,
+					       enum sec_c_mode mode,
+					       uint8_t *c_key_len)
+{
+	switch (mode) {
+	case C_MODE_ECB:
+	case C_MODE_CBC:
+	case C_MODE_CTR:
+		switch (key_len) {
+		case AES_KEYSIZE_128:
+			*c_key_len = CKEY_LEN_128_BIT;
+			break;
+		case AES_KEYSIZE_192:
+			*c_key_len = CKEY_LEN_192_BIT;
+			break;
+		case AES_KEYSIZE_256:
+			*c_key_len = CKEY_LEN_256_BIT;
+			break;
+		default:
+			EMSG("Invalid AES key size");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		break;
+	case C_MODE_XTS:
+		switch (key_len) {
+		case XTS_KEYSIZE_128:
+			*c_key_len = CKEY_LEN_128_BIT;
+			break;
+		case XTS_KEYSIZE_256:
+			*c_key_len = CKEY_LEN_256_BIT;
+			break;
+		default:
+			EMSG("Invalid AES-XTS key size");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		break;
+	default:
+		EMSG("Unsupported AES mode");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_cipher_sm4_get_c_key_len(size_t key_len,
+					       enum sec_c_mode mode,
+					       uint8_t *c_key_len)
+{
+	switch (mode) {
+	case C_MODE_ECB:
+	case C_MODE_CBC:
+	case C_MODE_CTR:
+		if (key_len != AES_KEYSIZE_128) {
+			EMSG("Invalid SM4 key size");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		*c_key_len = CKEY_LEN_128_BIT;
+		break;
+	case C_MODE_XTS:
+		if (key_len != XTS_KEYSIZE_128) {
+			EMSG("Invalid SM4-XTS key size");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		*c_key_len = CKEY_LEN_128_BIT;
+		break;
+	default:
+		EMSG("Unsupported SM4 mode");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_cipher_set_key(struct sec_cipher_ctx *c_ctx,
+				     uint8_t *key1, size_t key1_len,
+				     uint8_t *key2, size_t key2_len)
+{
+	size_t key_len = key1_len + key2_len;
+	TEE_Result ret = TEE_SUCCESS;
+	uint8_t c_key_len = 0;
+
+	switch (c_ctx->alg) {
+	case C_ALG_DES:
+		ret = sec_cipher_des_get_c_key_len(key_len, &c_key_len);
+		break;
+	case C_ALG_3DES:
+		ret = sec_cipher_3des_get_c_key_len(key_len, &c_key_len);
+		break;
+	case C_ALG_AES:
+		ret = sec_cipher_aes_get_c_key_len(key_len, c_ctx->mode,
+						   &c_key_len);
+		break;
+	case C_ALG_SM4:
+		ret = sec_cipher_sm4_get_c_key_len(key_len, c_ctx->mode,
+						   &c_key_len);
+		break;
+	default:
+		EMSG("Invalid cipher type %#"PRIx8, c_ctx->alg);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	if (ret)
+		return ret;
+
+	c_ctx->key_dma = virt_to_phys(c_ctx->key);
+	if (!c_ctx->key_dma) {
+		EMSG("c_key_dma is NULL");
+		return TEE_ERROR_GENERIC;
+	}
+
+	c_ctx->key_len = key_len;
+	c_ctx->c_key_len = c_key_len;
+
+	memcpy(c_ctx->key, key1, key1_len);
+	memcpy(c_ctx->key + key1_len, key2, key2_len);
+
+	return TEE_SUCCESS;
+}
+
+static void ctr_iv_inc(uint64_t *ctr, uint32_t inc)
+{
+	uint64_t v0 = TEE_U64_FROM_BIG_ENDIAN(ctr[0]);
+	uint64_t v1 = TEE_U64_FROM_BIG_ENDIAN(ctr[1]);
+
+	/* increment counter (128-bit int) by inc */
+	if (ADD_OVERFLOW(v1, inc, &v1))
+		v0++;
```

Comment:
The macro description states that `The macros return true if an overflow occurred and *res is undefined.`.
I think your use of the macro is fine and `v1` contains the target incremented 64b LSB value. I think the macro description should be updated. Otherwise, maybe prefer a local implementation like:
```c
if (v1 + inc < v1)
   v0++;
v1 += inc;

Reply:
Thank you for your review! The logic is clearer here with this macro, and` *res is undefined` doesn't happen here. Can I keep the use of the macro?

Reply:
I think you can.

Reply:
Can you give an `Acked-by: ...`? THX